### ISSUE TITLE
Stabilizing Iterator::intersperse breaks a large number of crates

### ISSUE THREAD

**Issue by Mark-Simulacrum** [rust-lang/rust#88967]
The 1.56 crater run shows that there are at least 59 root crates (or repositories) broken by this change, and 260 crates.io crates along with 278 github repositories which depend on those.

See https://crater-reports.s3.amazonaws.com/beta-1.56-2/index.html > "regressed: root results" > E0034 for the root crates.

cc some threads with partial discussion:

* https://rust-lang.zulipchat.com/#narrow/stream/241545-t-release/topic/possible.20semi-breaking.20change.20in.20next.20release
* https://github.com/rust-lang/rust/issues/88878

Nominating for T-libs-api discussion as the FCP on https://github.com/rust-lang/rust/issues/79524 seems to have been made without commentary about the expected level of breakage, so I suspect it was not discussed.

**Comment by the8472** [comment]
Given #48552, https://github.com/rust-itertools/itertools/pull/576 might solve this assuming it gets merged and released before 1.56 reaches stable.

**Comment by Mark-Simulacrum** [comment]
Dropping from the 1.56 milestone -- no longer release sensitive.

**Comment by yaahc** [comment]
@hamza1311 it causes ambiguous name resolution errors on files that already use the itertools intersperse method because now both the itertools and iterator traits are in scope and have a method with the same name, so the compiler goes from having one option to two options and produces an error because it can't pick arbitrarily from two options for which intersperse method to use.

**Comment by ckaran** [comment]
@yaahc wrote:
> The plan we arrived upon is to first revert the stabilization and then plan on stabilizing the feature in a future release, @Mark-Simulacrum suggested 1.65.

Is this still the plan?

**Comment by cdstanford** [comment]
If `.intersperse()` causes name resolution conflicts, the obvious thing to do is to rename. But why not also offer `.join()` which combines intersperse with collect (i.e., what users of this API actually want to do)?

**Comment by jswrenn** [comment]
Renaming is a somewhat unsatisfying solution. That approach sets us up to have our own little [smooshgate](https://developer.chrome.com/blog/smooshgate/) potentially every time a new method is added to the standard library. I *am* considering renaming every method in itertools to start with `it_`, but that'll only be a partial solution. A considerable number of itertools users aren't on the latest release. 

[RFC 2845](https://rust-lang.github.io/rfcs/2845-supertrait-item-shadowing.html) establishes a mechanism for reducing the breakage resulting from most of these situations, but requires implementation.

**Comment by cdstanford** [comment]
I don't see how it's analogous to smooshgate, because monkeypatching doesn't exist in Rust. itertools isn't using bad practice here.

There are plenty of reasonable alternative names like `.weave` or `.interleave_item`. How does choosing one of these set a bad precedent?

**Comment by the8472** [comment]
It is analogous in the sense that the respective standard library should not base its naming decisions on what downstream crates do. It's a violation of namespacing. That not fully qualified method resolution can lead to such conflicts is a known issue but it's better to resolve that rather than working around it.

**Comment by cdstanford** [comment]
I agree, but as a short-term solution renaming prevents this being blocked on what is a much bigger issue with Rust's namespacing. Otherwise it seems like this (and many similar stdlib improvements) are blocked for the near-term forseeable future, if I'm understanding the discussion correctly.

**Comment by the8472** [comment]
The RFC fixing the issue is already accepted and only needs to be implemented. Itertools provides that API for people that need it today, so there's no reason to rush.
